File: internal/anim/progress.go
// Package anim provides a spring-eased progress bar ported from
// the Rust donk-anim crate.
package anim

import (
	"math"
	"strconv"
	"strings"
)

// Progress is a spring-physics progress bar.
type Progress struct {
	Spring *Spring
	Width  int
	Label  string
}

// NewProgress creates a progress bar with default parameters.
func NewProgress() *Progress {
	return &Progress{
		Spring: NewSpring(),
		Width:  40,
		Label:  "progress",
	}
}

// SetWidth sets the minimum bar width.
func (p *Progress) SetWidth(width int) {
	if width > p.Width {
		p.Width = width
	}
	p.Spring.setWidth(p.Width)
}

// SetPercent sets the spring target as a 0-1 fraction.
func (p *Progress) SetPercent(pct float64) {
	if pct < 0 {
		pct = 0
	}
	if pct > 1 {
		pct = 1
	}
	p.Spring.Target = pct
}

// IncrPercent increments the spring target by delta.
func (p *Progress) IncrPercent(delta float64) {
	p.SetPercent(p.Spring.Target + delta)
}

// Poke toggles progress and injects spring velocity.
func (p *Progress) Poke() {
	p.Spring.Poke()
	if p.Spring.Target >= 1-math.Nextafter(1, 2) {
		p.Spring.Target = 0
		p.Spring.Pos = 0
		p.Spring.Vel = 0
	}
}

// Update advances the spring and frame counter.
func (p *Progress) Update() {
	p.Spring.Update()
}

// Percent returns the current spring position clamped to 0-1.
func (p *Progress) Percent() float64 {
	if p.Spring.Pos < 0 {
		return 0
	}
	if p.Spring.Pos > 1 {
		return 1
	}
	return p.Spring.Pos
}

// Render returns a text progress bar with label.
func (p *Progress) Render() string {
	w := p.Width
	if w > 48 {
		w = 48
	}
	fill := int(math.Round(p.Percent() * float64(w)))
	if fill > w {
		fill = w
	}
	bar := strings.Repeat("█", fill) + strings.Repeat("░", w-fill)
	pct := int(math.Round(p.Percent() * 100))
	return p.Label + "\n\n[" + bar + "] " + strconv.Itoa(pct) + "%\n\ntarget=" + strconv.FormatFloat(p.Spring.Target*100, 'f', 0, 64) + "% · harmonica spring · space +25%"
}

File: internal/ui/anim/library/rust-reference/progress.rs
//! Spring-eased progress bar (Bubble Tea progress-animated DNA).

use ratatui::prelude::*;

use crate::spring::Spring;

#[derive(Debug, Clone)]
pub struct Progress {
    spring: Spring,
    width: usize,
    label: String,
}

impl Default for Progress {
    fn default() -> Self {
        Self::new()
    }
}

impl Progress {
    pub fn new() -> Self {
        let mut spring = Spring::new();
        spring.target = 0.0;
        spring.pos = 0.0;
        Self {
            spring,
            width: 40,
            label: "progress".into(),
        }
    }

    pub fn set_width(&mut self, width: usize) {
        self.width = width.max(12);
        self.spring.set_width(self.width);
    }

    pub fn set_percent(&mut self, pct: f64) {
        self.spring.target = pct.clamp(0.0, 1.0);
    }

    pub fn incr_percent(&mut self, delta: f64) {
        self.set_percent(self.spring.target + delta);
    }

    pub fn poke(&mut self) {
        self.incr_percent(0.25);
        if self.spring.target >= 1.0 - f64::EPSILON {
            self.spring.target = 0.0;
            self.spring.pos = 0.0;
            self.spring.vel = 0.0;
        }
    }

    pub fn update(&mut self) {
        self.spring.update();
    }

    pub fn percent(&self) -> f64 {
        self.spring.pos.clamp(0.0, 1.0)
    }

    pub fn view_lines(&self) -> Vec<Line<'static>> {
        let w = self.width.min(48);
        let fill = ((self.percent() * w as f64).round() as usize).min(w);
        let bar = format!("{}{}", "█".repeat(fill), "░".repeat(w - fill));
        let pct = (self.percent() * 100.0).round() as i32;
        vec![
            Line::from(Span::styled(
                self.label.clone(),
                Style::default().add_modifier(Modifier::BOLD),
            )),
            Line::from(""),
            Line::from(Span::styled(
                format!("[{bar}] {pct}%"),
                Style::default().fg(Color::Rgb(0x6B, 0x50, 0xFF)),
            )),
            Line::from(""),
            Line::from(format!(
                "target={:.0}% · harmonica spring · space +25%",
                self.spring.target * 100.0
            )),
        ]
    }
}
